commands/lib/search.go

```diff
@@ -38,24 +38,141 @@ func LibrarySearch(ctx context.Context, req *rpc.LibrarySearchRequest) (*rpc.Lib
+	res := lib.Name + " " +
+		lib.Latest.Paragraph + " " +
+		lib.Latest.Sentence + " " +
+		lib.Latest.Author + " "
+	for _, include := range lib.Latest.ProvidesIncludes {
+		res += include + " "
+	}
+	return res
+}
+
+// matcherFromQueryString returns a closure that takes a library as a
+// parameter and returns true if the library matches the query.
+func matcherFromQueryString(query string) func(*librariesindex.Library) bool {
+	// A qv-query is one using <qualifier>[:=]<value> syntax.
+	qvQuery := strings.Contains(query, ":") || strings.Contains(query, "=")
+
+	if !qvQuery {
+		queryTerms := utils.SearchTermsFromQueryString(query)
+		return func(lib *librariesindex.Library) bool {
+			return utils.Match(defaultLibraryMatchExtractor(lib), queryTerms)
+		}
+	}
+
+	joinedStrings := func(strs []string) string {
+		return strings.Join(strs, " ")
+	}
+
+	qualifiers := []struct {
+		key       string
+		extractor func(*librariesindex.Library) string
+	}{
+		// The library name comes from the Library object.
+		{"name", func(lib *librariesindex.Library) string { return lib.Name }},
+
+		// All other values come from the latest Release.
+		{"architectures", func(lib *librariesindex.Library) string { return joinedStrings(lib.Latest.Architectures) }},
+		{"author", func(lib *librariesindex.Library) string { return lib.Latest.Author }},
+		{"category", func(lib *librariesindex.Library) string { return lib.Latest.Category }},
+		{"dependencies", func(lib *librariesindex.Library) string {
+			names := []string{}
+			for _, dep := range lib.Latest.Dependencies {
+				names = append(names, dep.GetName())
+			}
+			return joinedStrings(names)
+		}},
+		{"maintainer", func(lib *librariesindex.Library) string { return lib.Latest.Maintainer }},
+		{"paragraph", func(lib *librariesindex.Library) string { return lib.Latest.Paragraph }},
+		{"sentence", func(lib *librariesindex.Library) string { return lib.Latest.Sentence }},
+		{"types", func(lib *librariesindex.Library) string { return joinedStrings(lib.Latest.Types) }},
+		{"version", func(lib *librariesindex.Library) string { return lib.Latest.Version.String() }},
+		{"website", func(lib *librariesindex.Library) string { return lib.Latest.Website }},
+	}
+
+	queryTerms := matcherTokensFromQueryString(query)
+
+	return func(lib *librariesindex.Library) bool {
+		matched := true
+		for _, term := range queryTerms {
+
+			// Flag indicating whether the search term matched a known qualifier
+			knownQualifier := false
+
+			for _, q := range qualifiers {
+				if strings.HasPrefix(term, q.key+":") {
+					target := strings.TrimPrefix(term, q.key+":")
+					matched = (matched && utils.Match(q.extractor(lib), []string{target}))
+					knownQualifier = true
+					break
+				} else if strings.HasPrefix(term, q.key+"=") {
+					target := strings.TrimPrefix(term, q.key+"=")
+					matched = (matched && strings.ToLower(q.extractor(lib)) == target)
+					knownQualifier = true
+					break
+				}
+			}
+
+			if !knownQualifier {
```

Comment:
With `qualifiers` being a map we can avoid this for loop. Changing how we extract the qualifier key from `term`: if the key exists in the map we take the associated extractor, otherwise, we fallback to the default one.